### PR TITLE
[Protocol3] Allow permissioned account creation through an IAddressWhitelist contract

### PR DESCRIPTION
**Pull request by dong77**
This is to address #317.
 The signature verification part has not been implemented yet. But please review the overall solution.

### CODE REVIEW

**Comment by dong77** [comment]
The first 64 bits can be a timestamp, the rest bits should be a multisig against `hash(user_address, timestamp)`

**Comment by Brechtpd** [comment]
Looks good.

Any reason you went for the signature only approach (with some flexibility using multisig, but still a bit limited)? I still like the idea of doing stuff like this in a contract with an interface like `isWhitelisted(address accountCreator, address accountOwner, bytes permission)` instead of trying to come up with a data structure/implementation in the Exchange itself that is hopefully flexible enough.

The DEX would just set some contract, just like `_permissionedAccountCreation` is set now (if the contract is `0x0` all account creations are allowed). `bytes memory permission` would be some data completely opaque to the Exchange contract. It seems like this would move complexity outside of the Exchange contract and would be extremely flexible (the contract can be changed at any time, the DEX can call methods on this contract so that some signature can be enabled/disabled, ...).

**Comment by dong77** [comment]
> Looks good.
> 
> Any reason you went for the signature only approach (with some flexibility using multisig, but still a bit limited)? I still like the idea of doing stuff like this in a contract with an interface like `isWhitelisted(address accountCreator, address accountOwner, bytes permission)` instead of trying to come up with a data structure/implementation in the Exchange itself that is hopefully flexible enough.
> 
> The DEX would just set some contract, just like `_permissionedAccountCreation` is set now (if the contract is `0x0` all account creations are allowed). `bytes memory permission` would be some data completely opaque to the Exchange contract. It seems like this would move complexity outside of the Exchange contract and would be extremely flexible (the contract can be changed at any time, the DEX can call methods on this contract so that some signature can be enabled/disabled, ...).

Will do that way.

**Comment by dong77** [comment]
@Brechtpd please take another look. I suspect this will break some tests. Any suggestion for me where I should start fixing tests?

**Comment by Brechtpd** [comment]
Looking good.

I would suggest fixing all updated functions in `testExchangeUtil.ts` first so pretty much all tests will work again.

**Comment by dong77** [comment]
@Brechtpd  ready for review.

File: packages/loopring_v3/contracts/iface/IExchange.sol [hunk 1]
```diff
@@ -200,12 +206,16 @@ contract IExchange
     /// @param  pubKeyX The first part of the account's trading EdDSA public key
     /// @param  pubKeyY The second part of the account's trading EdDSA public key.
     ///                 Note that pubkeyX and pubKeyY cannot be both `1`.
+    /// @param  permission The operator's signature to allow the creation of new accounts.
```
Comment (Brechtpd) [inline packages/loopring_v3/contracts/iface/IExchange.sol]: 'The operator's signature' should be some data used for whitelisting.

File: packages/loopring_v3/contracts/iface/IExchange.sol [hunk 2]
```diff
@@ -604,6 +614,9 @@ contract IExchange
     ///
     /// @param  pubKeyX The first part of the account's trading EdDSA public key
     /// @param  pubKeyY The second part of the account's trading EdDSA public key
+    /// @param  permission The operator's signature to allow the creation of new accounts,
```
Comment (Brechtpd) [inline packages/loopring_v3/contracts/iface/IExchange.sol]: Same here

**Pull request merged**